platform/engineering/design-docs/2022-09-06-mobile-filtering.md

```diff
@@ -0,0 +1,162 @@
+# Mobile Filtering
+
+**Author(s):** Kris Pethtel, Andrew Herzberg
+**Last Updated:** September 6, 2022
+**Status:** **Draft**
+**Approvers:** FE, BE
+
+## Overview
+
+The vets-api has a filter system built in to model collections that allows clients to send query params specifying which records they want to include in the response. For example, a user can add a param like, `filter[refillStatus][eq]=refillinprocess` to indicate that they only want prescriptions where refill_status equals refillinprocess. This is relatively easy to use on the back end, but it has a few flaws.
+
+Most notably, it requires that any filterable attributes be whitelisted in the model along with any operations that are permitted on that attribute. This is compounded by the fact that some of the models we use are defined outside of the mobile module, meaning that changing them requires outside approval.
+
+The filtering code itself has some shortcomings as well, such as poor handling of cases where the client wants to fetch records where an attribute is one of multiple values, and we can't easily change the filtering logic because that would impact other teams who use it.
+
+We would like to give the mobile client the ability to filter by any attributes on the model without adding explicit handling within the model. We only want to allow filtering on top-level attributes. In other words, if a client wants to filter appointments based on location street address, that would not be allowed because location street address is not top-level.
+
+Because we don't want to whitelist valid operations per attribute on the model, it's possible that some queries will not be possible or will not yield desireable results. For example, requests for equality on a hash object would not make sense because we do not intend to allow such complex operations. Such a request should result in a 422 response. The filters will need robust error handling and logging for cases that either don't make sense or yield errors.
+
+## How to Transmit Filter Options
+
+### Query Params
+#### Example
+```/v0/health/rx/prescriptions?filter[refill_status][eq]=refillinprocess```
+
+#### Pros
+  - This is how the existing filtering works, allowing for backwards compatibility.
+  - No versioning of endpoints required.
+  - We know it works and doesn't yield any big surprises.
+  - Vets-api currently implementation of this uses query params so some logic could be borrowed from this.
+  - Sticks with convention of other mobile endpoints.
+
+#### Cons
+  - Longer, hard to read urls for requests with multiple filter parameters. URL length limits depend on browswers and webservers. Browswers are not relevant for our purposes, and popular webservers like nginx and apache support at least 4000 characters. Encoding special characters used in the filter query params will use some of that space, but it's unlikely that the 4000 character limit will be a problem.
```

Comment:
I mentioned a 2048 character limit to Andrew and Jayson. That was a misunderstanding on my part. That limit exists in some web browsers, which has no impact on us. Web servers also have limits, but the lowest I could find that might be relevant to us was 4k characters. I tested >8k and it seemed to work but it's hard to be certain because most of the characters were unused junk.